src/libjson/json.c

```diff
@@ -0,0 +1,1217 @@
+
+    CHECK(_invoke_callback(parser, JSON_REASON_END_OBJECT, JSON_TYPE_NULL, NULL));
+
+    parser->depth--;
+
+done:
+    return result;
+}
+
+static json_result_t _get_number(
+    json_parser_t* parser,
+    json_type_t* type,
+    json_union_t* un)
+{
+    json_result_t result = JSON_OK;
+    char c;
+    int isInteger = 1;
+    char* end;
+    const char* start = parser->ptr;
+
+    /* Skip over any characters that can comprise a number */
+    while (parser->ptr != parser->end && _is_number_char(*parser->ptr))
+    {
+        c = *parser->ptr;
+        parser->ptr++;
+
+        if (_is_decimal_or_exponent(c))
+            isInteger = 0;
+    }
+
+    if (isInteger)
+    {
+        *type = JSON_TYPE_INTEGER;
+        un->integer = _strtol(start, &end, 10);
+    }
+    else
+    {
+        *type = JSON_TYPE_REAL;
+        un->real = _strtod(start, &end);
+    }
+
+    if (!end || end != parser->ptr)
+        RAISE(JSON_BAD_SYNTAX);
+
+done:
+    return result;
+}
+
+/* value = false / null / true / object / array / number / string */
+static json_result_t _get_value(json_parser_t* parser)
+{
+    json_result_t result = JSON_OK;
+    char c;
+
+    /* Skip whitespace */
+    while (parser->ptr != parser->end && _isspace(*parser->ptr))
+        parser->ptr++;
+
+    /* Fail if output exhausted */
+    if (parser->ptr == parser->end)
+        RAISE(JSON_EOF);
+
+    /* Read the next character */
+    c = _tolower(*parser->ptr++);
+
+    switch (c)
+    {
+        case 'f':
+        {
+            json_union_t un;
+
+            if (_expect(parser, JSON_STRLIT("alse")) != 0)
+                RAISE(JSON_BAD_SYNTAX);
+
+            un.boolean = 0;
+
+            CHECK(_invoke_callback(
+                parser,
+                JSON_REASON_VALUE,
+                JSON_TYPE_BOOLEAN,
+                &un));
+
+            break;
+        }
+        case 'n':
+        {
+            if (_expect(parser, JSON_STRLIT("ull")) != 0)
+                RAISE(JSON_BAD_SYNTAX);
+
+            CHECK(_invoke_callback(
+                parser,
+                JSON_REASON_VALUE,
+                JSON_TYPE_NULL,
+                NULL));
+
+            break;
+        }
+        case 't':
+        {
+            json_union_t un;
+
+            if (_expect(parser, JSON_STRLIT("rue")) != 0)
+                RAISE(JSON_BAD_SYNTAX);
+
+            un.boolean = 1;
+
+            CHECK(_invoke_callback(
+                parser,
+                JSON_REASON_VALUE,
+                JSON_TYPE_BOOLEAN,
+                &un));
+
+            break;
+        }
+        case '{':
+        {
+            CHECK(_get_object(parser));
+            break;
+        }
+        case '[':
+        {
+            CHECK(_invoke_callback(
+                parser,
+                JSON_REASON_BEGIN_ARRAY,
+                JSON_TYPE_NULL,
+                NULL));
+
+            if (_get_array(parser) != JSON_OK)
+                RAISE(JSON_BAD_SYNTAX);
+
+            CHECK(_invoke_callback(
```

Comment:
It would be awesome of the parser did a forward-scan to determine the length of the array, but that may be more effort than we're willing to invest.

Reply:
Good suggestion. Done. The paths array is now called nodes and includes the name of the element, the size of the array (or zero if none), the index of the array (or zero if none).

Reply:
Awesome, thanks so much, this will make the client code much easier to write and read!